### PR TITLE
Add general preconditioner for Chebyshev

### LINKED ISSUE #3223: Add a class to represent a DiagonalMatrix
in reference to the discussion in https://github.com/dealii/dealii/pull/3221

@kronbichler 

>  Should we then try to add a minimal matrix interface to a matrix constructed from a vector with a template on the vector class? We could even think of re-designing the Chebyshev smoother to take a general Preconditioner operator as an argument. Besides offering the possibility to move the code more closely to a common smoother interface as discussed in #3184, this will also allow to use Chebyshev with something else than Jacobi, say block-Jacobi in DG methods. (I've already tested such methods and they are quite nice in some contexts.)

@tjhei 

> I agree that a DiagonalMatrix would be the way to go. But if nobody needs it right now, I think we shouldn't bother implementing it. I was also thinking that you might do different variants of this: lumping entries, only taking the diagonal, or something similar.

@bangerth 

> FWIW, I also agree that representing a diagonal matrix with a vector is a poor idea. If we need a diagonal matrix, let's have a class whose name and interface reflects that.

### PR DESCRIPTION
**Pull request by kronbichler**
This makes `PreconditionChebyshev` more flexible by allowing to pass general preconditioner classes to it. The old variant with the inverse diagonal is still the default choice, so there should be no change in functionality. One thing I had to change was that I had to move the initialization of the eigenvalues to the first vmult function because I do not have a vector exemplar any more in the initialize function. This changes output in some tests that print the iterations to the log stream but do not change the functionality otherwise.

Fixes #3223.

Relates to #2534.

### CODE REVIEW

File: include/deal.II/lac/diagonal_matrix.h [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+   * Number of rows of this matrix. To remember: this matrix is an <i>m x
+   * m</i>-matrix.
+   */
+  size_type m () const;
+
+  /**
+   * Number of columns of this matrix. To remember: this matrix is an <i>n x
+   * n</i>-matrix.
+   *
+   * Depending on the layout of the underlying vector, only some indices can
```
Comment (davydden) [inline include/deal.II/lac/diagonal_matrix.h]: that comment is probably out of place.

File: include/deal.II/lac/diagonal_matrix.h [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+  value_type &operator()(size_type i, size_type j);
+
+  /**
+   * Add an array of values given by <tt>values</tt> in the given global
+   * matrix row at columns specified by col_indices. Due to the storage of
+   * this matrix, entries are only added to the diagonal of the matrix. All
+   * other entries are ignored and no exception is thrown.
+   *
+   * This function is for a consistent interface with the other matrix classes
+   * in deal.II.
```
Comment (davydden) [inline include/deal.II/lac/diagonal_matrix.h]: maybe add that it's indent to be used with `constraints.distributed_local_to_global()` to get exactly the same diagonal as in the sparse matrix.

File: include/deal.II/lac/precondition.h [hunk 3]
```diff
@@ -811,19 +811,54 @@ class PreconditionPSOR : public PreconditionRelaxation<MatrixType>
...
+ * class can also be used as a direct solver.
+ *
+ * <h4>Requirements on templated classes</h4>
+ *
+ * For use with this class, the class MatrixType must be derived from
+ * Subscriptor because a SmartPointer to MatrixType is held in the class. In
+ * particular, this means that the matrix object needs to persist during the
+ * lifetime of PreconditionChebyshev. Regarding the preconditioner class, the
+ * object needs to be default-constructible and copy-constructible. The
+ * variable is held in a variable of type AdditionalData that is kept inside
```
Comment (davydden) [inline include/deal.II/lac/precondition.h]: maybe reword

> the variable is held in a variable...

File: include/deal.II/lac/precondition.h [hunk 4]
```diff
@@ -811,19 +811,54 @@ class PreconditionPSOR : public PreconditionRelaxation<MatrixType>
...
+ * the class.
+ *
+ * @author Martin Kronbichler, 2009, 2016; extension for full compatibility with
  * LinearOperator class: Jean-Paul Pelteret, 2015
  */
-template <typename MatrixType=SparseMatrix<double>, class VectorType=Vector<double> >
+template <typename MatrixType=SparseMatrix<double>, typename VectorType=Vector<double>, typename PreconditionerType=DiagonalMatrix<VectorType> >
```
Comment (davydden) [inline include/deal.II/lac/precondition.h]: why `PreconditionerType` is `DiagonalMatrix<>`? I mean, it's not really a preconditioner, it's a class to store the diagonal. Maybe rename this template parameter to something like `DiagonalType`?

File: include/deal.II/lac/precondition.h [hunk 5]
```diff
@@ -905,8 +940,15 @@ class PreconditionChebyshev : public Subscriptor
 
     /**
      * Stores the inverse of the diagonal of the underlying matrix.
+     *
+     * @deprecated Set the variable preconditioner instead.
```
Comment (davydden) [inline include/deal.II/lac/precondition.h]: supposedly you mean template parameter, no the variable?

File: include/deal.II/lac/precondition.h [hunk 6]
```diff
@@ -905,8 +940,15 @@ class PreconditionChebyshev : public Subscriptor
...
      */
     VectorType matrix_diagonal_inverse;
+
+    /**
+     * Stores the inverse of the diagonal of the underlying matrix.
```
Comment (davydden) [inline include/deal.II/lac/precondition.h]: That's not inverse, or is it?

File: tests/lac/diagonal_matrix.cc [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+// Tests DiagonalMatrix
```
Comment (davydden) [inline tests/lac/diagonal_matrix.cc]: could you please add one more which would be closer to the intended usage in assembly, i.e. compare `SparseMatrix<double>` to `DiagonalMatrix` for `FE_Q(2)` and triangulation with hanging nodes? (laplace or mass matrix, does not matter).

File: include/deal.II/lac/precondition.h (hunk 4 above)
Comment (davydden) [inline include/deal.II/lac/precondition.h]: or `DiagonalStorage` or something along these lines.

Comment (davydden) [inline include/deal.II/lac/precondition.h]: looking through the code, i see why you called it `PreconditionerType` as you need `vmult` of that class. How about `DiagonalOperator`?

Comment (davydden) [inline include/deal.II/lac/precondition.h]: from the usage i think it's `InverseDiagonalOperator`, right?

File: include/deal.II/lac/precondition.h [hunk 8]
```diff
@@ -1738,61 +1796,82 @@ namespace internal
...
+    void
+    initialize_preconditioner(const MatrixType   &matrix,
+                              PreconditionerType &preconditioner,
+                              VectorType         &diagonal_inverse)
     {
-      DiagonalPreconditioner (const VectorType &vector)
-        :
-        diagonal_vector(vector)
-      {}
+      AssertDimension(preconditioner.m(), matrix.m());
```
Comment (davydden) [inline include/deal.II/lac/precondition.h]: looks like a dummy `initialize_preconditioner()` which does not do anything, should there be 
```
AssertThrow(false, ExcNotImplemented())
```
?

File: include/deal.II/lac/diagonal_matrix.h (hunk 1 above)
Comment (kronbichler) [inline include/deal.II/lac/diagonal_matrix.h]: Oh yes, that is a copy-paste artifact that should actually have gone to two comments below.

File: include/deal.II/lac/diagonal_matrix.h (hunk 2 above)
Comment (kronbichler) [inline include/deal.II/lac/diagonal_matrix.h]: That's a good point. I will add this.

Comment (kronbichler) [inline include/deal.II/lac/diagonal_matrix.h]: In case one uses this matrix in `distribute_local_to_global`, one needs to call `compress(VectorOperation::add)` on the vector. Since the class is so closely linked to the vector (via the template) my opinion is that it is enough to expose the underlying vector that can call `compress`.

File: include/deal.II/lac/precondition.h (hunk 4 above)
Comment (kronbichler) [inline include/deal.II/lac/precondition.h]: @davydden The preconditioner type is really general and needs not only be a diagonal matrix. The diagonal matrix is just the simplest example and the default for backward compatibilty. Given how much has changed I'm quite proud of having managed to make this change without breaking backward compatibility and still going through the fast path with optimized vector operations for the diagonal matrix case. (These operations are on the critical path of computations that run on 10k cores for weeks, so the ~5% run time saved by this optimization is really worth it in my applications.)

I will try to create another test where the preconditioner is actually an ILU of the matrix. The motivation for the whole extension of this class is that I want to use block-Jacobi preconditioner objects inside Chebyshev, not just those with the diagonal.

File: include/deal.II/lac/precondition.h (hunk 6 above)
Comment (kronbichler) [inline include/deal.II/lac/precondition.h]: That comment is misleading. Will be updated.

File: include/deal.II/lac/precondition.h (hunk 8 above)
Comment (kronbichler) [inline include/deal.II/lac/precondition.h]: This function should do nothing because it is the case where we have a `PreconditionerType` object that is not just a diagonal matrix derived from a vector, so the user must have provided a suitable operation. The other `initialize_preconditioner` function is the one that links a vector with the inverse diagonal to the actual `DiagonalMatrix<VectorType>` variable, including the case where we access the matrix entries through `el(i,i)`.

**Comment by kronbichler** [comment]
This PR must be merged after #3296 as there is a conflict in `precondition.h` that is best resolved through this PR.

File: tests/lac/diagonal_matrix.cc (hunk 7 above)
Comment (kronbichler) [inline tests/lac/diagonal_matrix.cc]: We should add the `DiagonalMatrix<VectorType>` to the instantiation list of constraint matrix which will let us test this behavior. Do you want me to include this in this PR or should we create a separate one?

File: include/deal.II/lac/precondition.h (hunk 4 above)
Comment (davydden) [inline include/deal.II/lac/precondition.h]: I see, maybe mention that motivation in the class description?

File: tests/lac/diagonal_matrix.cc (hunk 7 above)
Comment (davydden) [inline tests/lac/diagonal_matrix.cc]: you are right, let's do it in a separate PR.

File: include/deal.II/lac/precondition.h (hunk 8 above)
Comment (davydden) [inline include/deal.II/lac/precondition.h]: thanks for explanation.

**Comment by kronbichler** [comment]
When trying to add a SparseILU as preconditioner, I realized that we cannot use that because it is not copy-constructible. I think that it would make sense to make the new `preconditioner` variable a `std_cxx11::shared_ptr<PreconditionerType>` so we can safely copy it. Any opinions?

**Comment by davydden** [comment]
do you need to share it, it could be `unique_ptr` otherwise.

**Comment by kronbichler** [comment]
I prefer to make it `shared_ptr` because otherwise I can't make the `AdditionalData` object passed to `initialize` const, which is our usual convention otherwise.

**Comment by davydden** [comment]
`shared_ptr` is fine with me.

**Comment by kronbichler** [comment]
This is now ready for re-review.

File: include/deal.II/lac/diagonal_matrix.h [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+   * Performs a transpose matrix-vector multiplication with the given
+   * matrix. Since this represents a diagonal matrix, exactly the same as
+   * vmult().
+   */
+  void Tvmult (VectorType       &dst,
+               const VectorType &src) const;
+
+  /**
+   * Adds the result of a matrix-vector multiplication into the destination
+   * vector dst. Needs to create a temporary vector, therefore it is not
```
Comment (davydden) [inline include/deal.II/lac/diagonal_matrix.h]: maybe have a mutable vector for this purpose so that you don't create it here every time, but only create it once in the constructor?

Comment (davydden) [inline include/deal.II/lac/diagonal_matrix.h]: or a pointer, which is innitialized first time `vmult_add` is called.

Comment (kronbichler) [inline include/deal.II/lac/diagonal_matrix.h]: I don't know how often `vmult_add` will be used. I don't think very often. But I'm fine to add a vector to the class  and initialize it in the first `vmult_add` call if someone sees an application that indeed does..

Comment (davydden) [inline include/deal.II/lac/diagonal_matrix.h]: up to you, most likely won't be used much, but on another hand it is also a trivial change.

**Comment by masterleinad** [comment]
Looks good to me!

File: include/deal.II/lac/diagonal_matrix.h [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+  VectorType &get_vector();
+
+  /**
+   * Returns a read-only reference to the underlying vector.
+   */
+  const VectorType &get_vector() const;
+
+  /**
+   * Number of rows of this matrix. To remember: this matrix is an <i>m x
+   * m</i>-matrix.
```
Comment (bangerth) [inline include/deal.II/lac/diagonal_matrix.h]: I think it would be useful to say that `m` (and `n` in the next function) is the size of the given vector of diagonal elements. 

The grammar of "To remember:" is a bit odd -- but I know you only copied that from elsewhere and I also know that the person who originally came up with this phrase is likely me, so I'll let that one slide ;-)

File: include/deal.II/lac/diagonal_matrix.h [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+  /**
+   * Read-only access to a value. This is restricted to the case where
+   * <i>i==j</i>.
+   *
+   * If the underlying vector is distributed with MPI, not all of the indices
+   * <i>i</i> might actually be accessible. Refer to the method
+   * <code>get_vector().locally_owned_elements()</code> for the entries that
+   * actually are accessible.
+   */
+  value_type operator()(size_type i, size_type j) const;
```
Comment (bangerth) [inline include/deal.II/lac/diagonal_matrix.h]: can you break the two arguments onto separate lines and make them `const`? Same for the next function.

File: include/deal.II/lac/diagonal_matrix.h [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+   * <code>get_vector().locally_owned_elements()</code> for the entries that
+   * actually are accessible.
+   */
+  value_type &operator()(size_type i, size_type j);
+
+  /**
+   * Add an array of values given by <tt>values</tt> in the given global
+   * matrix row at columns specified by col_indices. Due to the storage of
+   * this matrix, entries are only added to the diagonal of the matrix. All
+   * other entries are ignored and no exception is thrown.
```
Comment (bangerth) [inline include/deal.II/lac/diagonal_matrix.h]: This is interesting semantics. Out of curiosity: Why do you do it that way? Why *not* throw an exception if an entry to be added is non-zero?

File: include/deal.II/lac/diagonal_matrix.h (hunk 9 above)
Comment (bangerth) [inline include/deal.II/lac/diagonal_matrix.h]: I don't think it's that trivial. The function is `const`, so your temp vector needs to be `mutable`, which in turn implies that you need a mutex to make it thread-safe.

File: include/deal.II/lac/precondition.h [hunk 13]
```diff
@@ -905,8 +959,15 @@ class PreconditionChebyshev : public Subscriptor
 
     /**
      * Stores the inverse of the diagonal of the underlying matrix.
+     *
+     * @deprecated Set the variable @p preconditioner defined below instead.
      */
     VectorType matrix_diagonal_inverse;
```
Comment (bangerth) [inline include/deal.II/lac/precondition.h]: Are you actually still using this variable?

What happens if you mark this variable as `DEAL_II_DEPRECATED`?

File: include/deal.II/lac/precondition.h [hunk 14]
```diff
@@ -986,7 +1047,13 @@ class PreconditionChebyshev : public Subscriptor
   mutable VectorType update2;
 
   /**
-   * Stores the additional data provided to the initialize function.
+   * Internal vector used for the <tt>vmult</tt> operation.
+   */
+  mutable VectorType update3;
```
Comment (bangerth) [inline include/deal.II/lac/precondition.h]: You probably want to guard this mutable variable with a mutex, to make sure that one can call all `const` function more than once in parallel. I don't know whether that's going to be a common use case, but it's so incredibly hard to find these kinds of bugs that I think we do well to *always* guard mutable variables with a mutex.

File: include/deal.II/lac/diagonal_matrix.h (hunk 12 above)
Comment (davydden) [inline include/deal.II/lac/diagonal_matrix.h]: i think we need that for the following reason: When you assemble a matrix with hanging nodes, you still need to have a full local matrix to get correct diagonals. Since we will use the same code path in `constraints.distribute_local_to_globa()` with this matrix object, then we should not assume that this function will be given non-zeroes on diagonal only.

File: include/deal.II/lac/diagonal_matrix.h (hunk 9 above)
Comment (davydden) [inline include/deal.II/lac/diagonal_matrix.h]: fair point, i was not aware of this.

Comment (kronbichler) [inline include/deal.II/lac/diagonal_matrix.h]: We can postpone this for now and add it in case we see that it hurts.

File: include/deal.II/lac/diagonal_matrix.h (hunk 10 above)
Comment (kronbichler) [inline include/deal.II/lac/diagonal_matrix.h]: I will fix this documentation.

File: include/deal.II/lac/diagonal_matrix.h (hunk 12 above)
Comment (kronbichler) [inline include/deal.II/lac/diagonal_matrix.h]: Exactly, in the context of `ConstraintMatrix::distribute_local_to_global` where this might be used, we do not want to write separate code paths. Even if a user only passes a diagonal matrix to that function, constraints will create non-diagonal contributions nonetheless, so we are forced to skip non-diagonal entries rather than assert at this stage. Unless we see a good alternative to this, I'd keep the semantics like this. (Either I or @davydden will add that class to ConstraintMatrix in a later PR.)

File: include/deal.II/lac/precondition.h (hunk 14 above)
Comment (kronbichler) [inline include/deal.II/lac/precondition.h]: I will add a mutex inside the whole `vmult` function because the same applies to `update1` and `update2`. This class is simply not designed to be used in parallel - one needs to provide several such objects in case one wants to do that.

**Comment by kronbichler** [comment]
Except for the question regarding the `add` function of `DiagonalMatrix` and the deprecation warning where I do not really know what to do, this should be good to go.

File: include/deal.II/lac/diagonal_matrix.h (hunk 12 above)
Comment (bangerth) [inline include/deal.II/lac/diagonal_matrix.h]: OK. It just struck me as potentially surprising if we throw away entries -- because you get a different matrix than you were expecting. But I can live with it.

File: include/deal.II/lac/precondition.h (hunk 13 above)
Comment (bangerth) [inline include/deal.II/lac/precondition.h]: There is this `#pragma push/pop` thing (hidden behind a macro) we use to silence warnings in headers. Presumably we could do the same within regular code, but I've never used it. It may be worth a try, though.

**Comment by bangerth** [comment]
OK from my end unless you want to try to play with the warnings `#pragma` for a minute.

**Comment by kronbichler** [comment]
@bangerth My last commit introduces the pragmas (I didn't know that they also disable the deprecation warnings). I had to insert those statements in three places but I guess it's fine now.

**Comment by bangerth** [comment]
OK, thanks!

**Comment by bangerth** [comment]
Ah, I keep forgetting to ask for a squash :-( Oh well...

**Comment by davydden** [comment]
@bangerth apparently you can also squash yourself via GUI by using the drop-down menu near the green button for merging.

**Comment by bangerth** [comment]
Yes, I knw, but I think that's unfriendly to the author. But the point is that I just keep forgetting that one should squash...

**Comment by davydden** [comment]
> but I think that's unfriendly to the author.

fully agree.

**Comment by kronbichler** [comment]
Actually, I had tried to group my commits so one can separate the diagonal matrix commits from the Chebyshev commits (I forgot to merge the tests in the end) and also intentionally kept the last commit regarding deprecation separate in case that goes wrong on some platforms.

**Comment by bangerth** [comment]
Oh, ok, great. That's basically what I would have asked you to do -- to group things so that they logically build on each other and could, if necessary, be undone individually. Thanks for being proactive!
